torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -414,6 +415,59 @@ def inner(*flat_args):
                 is_result_of_custom_autograd_fn = True
             if isinstance(grad_fn, torch.autograd.function.BackwardCFunction):
                 is_result_of_custom_autograd_fn = True
+            # Whether we should save this output's FunctionalTensor or not.
+            #
+            # This will be used at runtime for reconstructing output views from
+            # their respective base tensors.
+            #
+            # The FunctionalTensor will be saved iff:
+            #
+            #   1. The output is an alias of some kind. Specifically, if the
+            #      output_type is either of:
+            #      (i) alias_of_input;
+            #      (ii) alias_of_intermediate;
+            #      (iii) alias_of_intermediate_save_as_output; or
+            #      (iv) alias_of_intermediate_base_is_user_output.
+            #
+            #   2. The base tensor of the output has not gone through a metadata
+            #      mutation. With this, we explicitly not support in-place view
+            #      operations.
+            should_save_functional_tensor = False
+
+            # Check if the base tensor has gone through a metadata mutation, and
+            # change should_save_functional_tensor appropriately.
+            def check_should_save_functional_tensor(f_arg, arg=None):
+                nonlocal should_save_functional_tensor
+
+                if not (
+                    isinstance(o, FunctionalTensor)
+                    and isinstance(f_arg, FunctionalTensor)
+                ):
+                    # Bail if we are not dealing with FunctionalTensor.
+                    return
+
+                if arg is None:
```

Comment:
I believe this check can be simplified. If I'm following correctly, the logic in `check_should_save_functional_tensor` is basically trying to figure out when there were any inplace views, making it unsafe to replay the view chain.

`inplace_views` are **only** a problem though for inputs (because they have a user-visible side-effect). Any inplace-views done on graph intermediates can be fully functionalized away into ordinary views.

Somethign like
```
if output_type in [AliasOfIntermediate, AliasOfIntermediateSaveAsOutput, AliasOfIntermedateBaseIsOutput]:
    functional_tensor = FunctionalTensorMetadataEq(o.elem)
    # you query the `InputAliasInfo` of the input to see if there was a metadata mutation,
    # instead of recomputing it directly
elif output_type is AliasOfInput and not input_info[base_idx].mutates_metadata
    functional_tensor = FunctionalTensorMetadataEq(o.elem)
else:
    functional_tensor = None
```

Reply:
Ah, I see. Since we are eliminating tensor mutation, we have access to the actual base tensor even after an inplace view operation. Thanks, Brian.